### ISSUE TITLE
Game windows managed by river report inverted aspect ratios/resolutions on some dual-monitor setups

### ISSUE THREAD

**Issue by vide0hanz** [riverwm/river#985]
This one is weird, and while doing research for this report I discovered a workaround that appears to resolve it (shared below), but I feel it may be worth reporting regardless as it seems related to #391. Hopefully it will at least help others searching the tracker for leads to the issue.

I have a dual monitor setup configured in my init like so:

```
# Set screen positions - DP-1 (landscape) to the right of DP-2 (portrait)
wlr-randr --output DP-1 --preferred --pos 1440,740 --adaptive-sync enabled
wlr-randr --output DP-2 --preferred --pos 0,0 --transform 90 --adaptive-sync enabled
```
And here is an example output of my preferred mode according to `wlr-randr` for both displays:

```
Modes:
    2560x1440 px, 164.834000 Hz (preferred, current)
    ...
```

The position is adjusted so that I can seamlessly move my mouse cursor into the secondary (DP-2) display in a way that feels natural. This works mostly as intended except many games running under xwayland (ie Steam games) get confused while using river compared to other compositors. 

## The problem:
Games will always report the following examples with the above configuration:

1. 1440x2560 [9:16]
2. 1080x1920 [9:16]
3.  800x1280 [10:16]

- At best, the game reports these inverted resolutions but is able to cope with displaying it and works fine.
- On average, the game appears to display correctly but has no clue what resolution it's running in and approximates it, and we get issues where clickable elements do not align with the cursor position, making window interaction challenging.
    - This can usually be worked around by telling the game to run in windowed mode if possible, and have river handle the fullscreen request.
- At worst, the game attempts to run at the requested resolution and is incredibly distorted/stretched.
    - This can sometimes be worked around by running in windowed mode. 

## Expected behavior:
The opposite of the problem :) -- aspect ratios should be correctly reported relative to the output the views are being displayed in, regardless of their position. Hyprland, for example, handles this correctly.

## Known workarounds:
1. Invert the positioning, ie:
```
# DP-1 at position 0,0 with DP-2 to the left
wlr-randr --output DP-1 --preferred --pos 0,0 --adaptive-sync enabled
wlr-randr --output DP-2 --preferred --pos -1440,-740 --transform 90 --adaptive-sync enabled
```
This somewhat solves the issue in the sense that the aspect ratios and resolutions are now correct, but then the behavior in #391 occurs, ie if Steam is launched on the secondary display in this example, it's negative positioning will not allow us to interact with it.

2. Simply setting the positioning of my secondary display to `-1,-1` appears to mitigate both issues, ie:
```
wlr-randr --output DP-1 --preferred --pos 1440,740 --adaptive-sync enabled
wlr-randr --output DP-2 --preferred --pos -1,-1 --transform 90 --adaptive-sync enabled
```

With workaround 2, aspect ratios are reported correctly in games and other xwayland views (steam is the only one I've tested) can still be interacted with. I am guessing this works because a 1px difference is too tiny to detect?

Whether or not this is truly a bug with river I'm not sure, but since this doesn't appear to be a widely reported issue with other wayland compositors I am posting it here regardless. Hopefully it at least helps others searching for solutions, if nothing else :)

**Comment by ifreund** [comment]
Since running `xrandr` properly reports two monitors with the correct dimensions/orientations when I test with a monitor configuration identical to yours I'm not sure what river could do to improve this. It seems like a bug/limitation in either Xwayland or possible wine (assuming you are running game through wine/proton/similar).

Perhaps programs are simply using the dimensions of the first or last output advertised to them and ignoring the others, I'm not sure if there's anything that can be done on river's side to fix this properly. At best, a workaround could potentially be added but as far as I can tell river is already behaving correctly and the problem lies elsewhere.

**Comment by vide0hanz** [comment]
True, I've only tested and noticed this behavior with wine/proton but it still is odd that so far river seems to be the only one that I can recall not reporting the correct aspect ratios, etc. I won't have time until the weekend, but I can test and report back what I find with a few other tiling wayland compositors if that would be helpful at all?

In the meantime I can try to reproduce with native titles and compare results when running under xwayland vs wayland.

**Comment by vide0hanz** [comment]
> Perhaps programs are simply using the dimensions of the first or last output advertised to them and ignoring the others, I'm not sure if there's anything that can be done on river's side to fix this properly. At best, a workaround could potentially be added but as far as I can tell river is already behaving correctly and the problem lies elsewhere.

You're probably on to something here. I just briefly tested this with the linux native build of Half Life and confirmed it was using xwayland with `xeyes`; while the window didn't appear where I expected it to appear, the resolution was reported correctly.

When I re-ran the same test using WINE/proton the resolution was reported as 1440x2560, which would be technically correct if it is indeed using the dimensions of the first reported output advertised to it, like you said. Even when I attempted to fullscreen the window, the viewport only allowed for 1440x2560 regardless (massive black bars on my landscape display)

I'm not entirely sure why setting the transformed display position to -1,-1 makes these views work as expected, but I'm personally fine doing that as a workaround. 

One thing that immediately comes to mind that river might be able to do as a workaround though (and please correct me if I am wrong), is to force views to inherit the output properties they're instructed to start on. For example, something like `riverctl rule-add -app-id "steam_app_*" output DP-1` works as you'd expect it to, but perhaps there could be another check to query the properties of what size/position DP-1 is and ensure new views get this information. Is that even viable??

**Comment by ifreund** [comment]
> One thing that immediately comes to mind that river might be able to do as a workaround though (and please correct me if I am wrong), is to force views to inherit the output properties they're instructed to start on. For example, something like riverctl rule-add -app-id "steam_app_*" output DP-1 works as you'd expect it to, but perhaps there could be another check to query the properties of what size/position DP-1 is and ensure new views get this information. Is that even viable??

River exposes information about all outputs to the xwayland process. How xwayland communicates that information to X11 clients is outside of river's control and I don't know the details of the X11 protocol bits involved.

Given the current information in this thread, I believe this to be an issue with upstream xwayland or wine. I wonder if the new wayland backend in upstream wine still has this issue, taking xwayland and X11 out of the equation seems like it could only make this situation better.

**Comment by vide0hanz** [comment]
I agree, seems specific to wine based on my observations. Perhaps the Wayland backend in wine will clear this up when that lands.

In the meantime, the workaround I mentioned appears to resolve this and I haven't noticed any weirdness with using it. Do you think it's worth adding this information to the wiki/FAQ section?

**Comment by ifreund** [comment]
> Do you think it's worth adding this information to the wiki/FAQ section?

I'd rather not as I don't fully understand how/why it works and don't want something that could potentially have additional, negative effects listed as a recommendation in the docs.
